include/boost/hana/experimental/printable.hpp

```diff
@@ -111,6 +111,18 @@ BOOST_HANA_NAMESPACE_BEGIN namespace experimental {
         }
     };
 
+    // model for hana::optional
+    template <>
+    struct print_impl<hana::optional_tag> {
+        template <typename O>
+        static auto apply(O const& optional) {
```

Comment:
When the optional is empty, you return `"nothing"`. Hence the return type will be deduced to be `char*` (or `char const*`, not sure), which is wrong. To make sure that we're returning a `std::string`, you could specify the return type as `std::string`. Or perhaps that was done on purpose?